lib/ReplaceOpenCLBuiltinPass.cpp

```diff
@@ -2562,128 +2565,113 @@ bool ReplaceOpenCLBuiltinPass::replaceHadd(Function &F, bool is_signed,
         if (auto vec_ty = dyn_cast<VectorType>(ty)) {
           min = ConstantVector::getSplat(vec_ty->getElementCount(), min);
           max = ConstantVector::getSplat(vec_ty->getElementCount(), max);
         }
-        auto zero = Constant::getNullValue(ty);
-        // Cannot add the nsw flag.
-        auto add = BinaryOperator::Create(Instruction::Add, op0, op1, "", Call);
-        auto add_gt_op0 = CmpInst::Create(Instruction::ICmp, CmpInst::ICMP_SGT,
-                                          add, op0, "", Call);
-        auto min_clamp = SelectInst::Create(add_gt_op0, min, add, "", Call);
-        auto add_lt_op0 = CmpInst::Create(Instruction::ICmp, CmpInst::ICMP_SLT,
-                                          add, op0, "", Call);
-        auto max_clamp = SelectInst::Create(add_lt_op0, max, add, "", Call);
-        auto op1_lt_0 = CmpInst::Create(Instruction::ICmp, CmpInst::ICMP_SLT,
-                                        op1, zero, "", Call);
-        result = SelectInst::Create(op1_lt_0, min_clamp, max_clamp, "", Call);
+        Value *op = nullptr;
+        if (is_add) {
+          op = builder.CreateAdd(a, b);
+        } else {
+          op = builder.CreateSub(a, b);
+        }
+        auto b_lt_0 = builder.CreateICmpSLT(b, Constant::getNullValue(ty));
+        auto op_gt_a = builder.CreateICmpSGT(op, a);
+        auto op_lt_a = builder.CreateICmpSLT(op, a);
+        auto neg_cmp = is_add ? op_gt_a : op_lt_a;
+        auto pos_cmp = is_add ? op_lt_a : op_gt_a;
+        auto neg_value = is_add ? min : max;
+        auto pos_value = is_add ? max : min;
+        auto neg_clamp = builder.CreateSelect(neg_cmp, neg_value, op);
+        auto pos_clamp = builder.CreateSelect(pos_cmp, pos_value, op);
+        return builder.CreateSelect(b_lt_0, neg_clamp, pos_clamp);
       }
     } else {
-      // Just use OpIAddCarry and use the carry to clamp the result.
-      auto ret_ty = GetPairStruct(ty);
-      auto add = clspv::InsertSPIRVOp(
-          Call, spv::OpIAddCarry, {Attribute::ReadNone}, ret_ty, {op0, op1});
-      auto ex0 = ExtractValueInst::Create(add, {0}, "", Call);
-      auto ex1 = ExtractValueInst::Create(add, {1}, "", Call);
-      auto cmp = CmpInst::Create(Instruction::ICmp, CmpInst::ICMP_EQ, ex1,
-                                 Constant::getNullValue(ty), "", Call);
-      result =
-          SelectInst::Create(cmp, ex0, Constant::getAllOnesValue(ty), "", Call);
-    }
-
-    return result;
+      // Replace with OpIAddCarry/OpISubBorrow and clamp to max/0 on a
+      // carr/borrow.
+      spv::Op op = is_add ? spv::OpIAddCarry : spv::OpISubBorrow;
+      auto clamp_value =
+          is_add ? Constant::getAllOnesValue(ty) : Constant::getNullValue(ty);
+      auto struct_ty = GetPairStruct(ty);
+      auto call =
+          InsertSPIRVOp(Call, op, {Attribute::ReadNone}, struct_ty, {a, b});
+      auto sub = builder.CreateExtractValue(call, {0});
+      auto borrow = builder.CreateExtractValue(call, {1});
```

Comment:
Nit: These variable names are specific to the `sub` case which tripped me up on a first skim. No big deal though.